src/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs

```diff
@@ -58,40 +59,133 @@ public void ConcatenatedGzipStreams()
+                {
+                    // read one byte in order to fill the inflater bufffer before copy
+                    Assert.Equal(3, gz.ReadByte());
+
+                    gz.CopyTo(decompressedData, copyToBufferSizeRequested);
+                    Assert.Equal(0, decompressedData.Length);
+                }
+
+                compressedStream.Seek(0, SeekOrigin.Begin);
+                using (Stream gz = new GZipStream(compressedStream, CompressionMode.Decompress, leaveOpen: true))
+                using (Stream decompressedData = new DerivedMemoryStream())
+                {
+                    // read one byte in order to fill the inflater bufffer before copy
+                    Assert.Equal(3, gz.ReadByte());
+
+                    await gz.CopyToAsync(decompressedData, copyToBufferSizeRequested);
+                    Assert.Equal(0, decompressedData.Length);
+                }
+            }
+        }
+
         [Theory]
-        [InlineData(1000, false, false, 1000, 1)]
-        [InlineData(1000, false, true, 0, 1)]
-        [InlineData(1000, true, false, 1000, 1)]
-        [InlineData(1000, false, false, 1, 1)]
-        [InlineData(1000, true, false, 1, 1)]
-        [InlineData(1000, false, false, 1001 * 24, 1)]
-        [InlineData(1000, true, false, 1001 * 24, 1)]
-        public async Task ManyConcatenatedGzipStreams(int streamCount, bool useAsync, bool useReadByte, int bufferSize, int bytesPerStream)
+        [InlineData(1000, TestScenario.Read, 1000, 1)]
+        [InlineData(1000, TestScenario.ReadByte, 0, 1)]
+        [InlineData(1000, TestScenario.ReadAsync, 1000, 1)]
+        [InlineData(1000, TestScenario.Copy, 1000, 1)]
+        [InlineData(1000, TestScenario.CopyAsync, 1000, 1)]
+        [InlineData(10, TestScenario.Read, 1000, 2000)]
+        [InlineData(10, TestScenario.ReadByte, 0, 2000)]
+        [InlineData(10, TestScenario.ReadAsync, 1000, 2000)]
+        [InlineData(10, TestScenario.Copy, 1000, 2000)]
+        [InlineData(10, TestScenario.CopyAsync, 1000, 2000)]
+        [InlineData(2, TestScenario.Copy, 1000, 0x2000-30)]
+        [InlineData(2, TestScenario.CopyAsync, 1000, 0x2000 - 30)]
+        [InlineData(1000, TestScenario.Read, 1, 1)]
+        [InlineData(1000, TestScenario.ReadAsync, 1, 1)]
+        [InlineData(1000, TestScenario.Read, 1001 * 24, 1)]
+        [InlineData(1000, TestScenario.ReadAsync, 1001 * 24, 1)]
+        [InlineData(1000, TestScenario.Copy, 1001 * 24, 1)]
+        [InlineData(1000, TestScenario.CopyAsync, 1001 * 24, 1)]
+        public async Task ManyConcatenatedGzipStreams(int streamCount, TestScenario scenario, int bufferSize, int bytesPerStream)
         {
-            await TestConcatenatedGzipStreams(streamCount, useAsync, useReadByte, bufferSize, bytesPerStream);
+            await TestConcatenatedGzipStreams(streamCount, scenario, bufferSize, bytesPerStream);
         }
 
         [Theory]
         [OuterLoop("Tests take a very long time to complete")]
```

Comment:
Non blocking question, just out of curiosity: Would it be beneficial to split this test into multiple Theory batches so they get executed concurrently and can finish sooner?

Reply:
Not sure how much we care about outerloop perf.  If you happen to run these and find out that could help then go ahead and do it.